wknml/nml_utils.py

```diff
@@ -0,0 +1,104 @@
+from math import sqrt, ceil
+from copy import deepcopy
+import networkx as nx
+import numpy as np
+
+from . import NML
+from .nml_generation import generate_graph, generate_nml
+
+
+def ensure_max_edge_length(nml_or_graph: Union[NML, nx.Graph], max_length: int) -> Union[NML, nx.Graph]:
+    # it is easier to operate on a graph
+    if isinstance(nml_or_graph, nx.Graph):
+        nml_graph = nml_or_graph
+    else:
+        nml_graph, parameter_dict = generate_graph(nml_or_graph)
+    max_id = detect_max_node_id_from_all_graphs(nml_graph)
+    next_valid_id = max_id + 1
+    for group in nml_graph.values():
+        for graph in group:
+            next_valid_id = ensure_max_edge_length_in_tree(graph, max_length, next_valid_id)
+
+    # return the same format as the input
+    if isinstance(nml_or_graph, nx.Graph):
+        return nml_graph
+    else:
+        return generate_nml(nml_graph, parameter_dict, globalize=False)
+
+
+def calculate_distance_between_nodes(node1: Dict, node2: Dict) -> float:
+    node1_position = node1['position']
+    node2_position = node2['position']
+    return sqrt((node1_position[0] - node2_position[0]) ** 2 +
+                (node1_position[1] - node2_position[1]) ** 2 +
+                (node1_position[2] - node2_position[2]) ** 2)
+
+
+def get_vector_between_nodes(node1: Dict, node2: Dict) -> List[int]:
+    node1_position = node1['position']
+    node2_position = node2['position']
+    return [node2_position[0] - node1_position[0],
+            node2_position[1] - node1_position[1],
+            node2_position[2] - node1_position[2]]
+
+
+def get_padding_node_position(node1: Dict, node2: Dict, relative_distance_along_vector: float) -> List[int]:
+    node1_position = node1['position']
+    vector_between_nodes = get_vector_between_nodes(node1, node2)
+    return [node1_position[0] + vector_between_nodes[0] * relative_distance_along_vector,
+            node1_position[1] + vector_between_nodes[1] * relative_distance_along_vector,
+            node1_position[2] + vector_between_nodes[2] * relative_distance_along_vector]
+
+
+def detect_max_node_id_from_all_graphs(graph_dict: Dict[str, nx.Graph]) -> int:
+    max_id = None
+    for group in graph_dict.values():
+        for tree in group:
+            max_id_of_current_tree = np.array(list(tree.nodes)).max()
+            max_id = max_id_of_current_tree if not max_id or max_id_of_current_tree > max_id else max_id
+
+    return max_id
+
+
+def ensure_max_edge_length_in_tree(graph: nx.Graph, max_length: int, current_id: int) -> int:
+    edges_to_be_added = []
+    edges_to_be_removed = []
+    nodes_to_be_added = []
+    for edge in graph.edges:
+        node1 = graph.nodes[edge[0]]
+        node2 = graph.nodes[edge[1]]
+        edge_distance = calculate_distance_between_nodes(node1, node2)
+        # add padding nodes if the distance is too high
+        if edge_distance > max_length:
+            number_of_padding_nodes = ceil(edge_distance / max_length)
```

Comment:
this is confusing as it is actually 
```suggestion
            number_of_padding_nodes = floor(edge_distance / max_length)
```

Reply:
I am sorry but I think this will not work.
Let the max_distance be 2 and the distance between the two vertices x, y be 8. Now you want exactly 3 padding nodes between them.
Now floor(8 / 2) = 4 and 4 + 1 = 5. Therefore the range will be [0, 1, 2, 3, 4]. We skip the first one. Therefore we would place [1, 2, 3, 4] - padding nodes, which is one too much. Actually the 4th node (the one that is too much) will have the exact same position as y. Therefore the last padding node is not needed.

Did I do any mistake? :worried:

Reply:
Yes, you are right! The point I am trying to make is about the naming of the variable. So I think, `number_of_padding_nodes` should be `ceil(edge_distance / max_length) -1`. Or you find a name which expresses the number of inserted nodes plus one.

Reply:
maybe`range_of_padding_nodes`?